Add HTTP auth check

drasko opened the pull request:
Signed-off-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>

drasko committed:
Add HTTP auth check

Signed-off-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>

mijicd commented on http/api/transport.go:
```diff
@@ -76,14 +77,32 @@ func decodeRequest(_ context.Context, r *http.Request) (interface{}, error) {
 	}
 
-	return apiKey, nil
+	identityURL := os.Getenv("HTTP_ADAPTER_MANAGER_URL") + "/identity"
```
This can be set at top-level, as a constant, instead of setting it per each request.

mijicd commented on http/api/transport.go:
```diff
@@ -76,14 +77,32 @@ func decodeRequest(_ context.Context, r *http.Request) (interface{}, error) {
+	c := &http.Client{}
+	resp, err := c.Do(req)
+	if err != nil {
+		return "", err
```
Make sure this error is properly handled. The safest way is to utilize domain errors.

drasko replied:
What do you mean by this? I thought that upper layer above `transport.go` handles this error, so I just propagate it.

mijicd replied:
Safely ignore this one, it was deprecated by #138 .

mijicd reviewed (changes requested):
Make sure the build is fixed.

drasko committed:
Add HTTP auth check

Signed-off-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>

drasko committed:
Add HTTP auth via Manager Client

Signed-off-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>

mijicd commented on http/api/transport.go:
```diff
@@ -76,14 +79,29 @@ func decodeRequest(_ context.Context, r *http.Request) (interface{}, error) {
 	}
 
-	return apiKey, nil
+	authURL := os.Getenv("HTTP_ADAPTER_MANAGER_URL")
```
It's better to perform this load & fallback at main, and just pass the configured value to this module.

drasko replied:
Sure, but how to pass to this module :)

mijicd replied:
Store the client reference to the service itself. Modify [this signature](https://github.com/mainflux/mainflux/blob/master/cmd/http/main.go#L47) and use it as an entrypoint for injecting the client. Then, store the client reference when initializing the transport layer through `MakeHandler` method and use it to authorize requests.

mijicd commented on http/api/transport.go:
```diff
@@ -76,14 +79,29 @@ func decodeRequest(_ context.Context, r *http.Request) (interface{}, error) {
-// TODO: contact an auth provider
 func authorize(r *http.Request) (string, error) {
 	var apiKey string
+	var id string
```
Not needed, it is created and initialized below.

mijicd committed:
Change the way manager client is injected

Instead of relying on modifying service structure, manager client is
injected through handler factory method, and store as a variable in that
module.

Signed-off-by: Dejan Mijic <dejan@mainflux.com>

mijicd committed:
Update dockerignore

Signed-off-by: Dejan Mijic <dejan@mainflux.com>
